### PR TITLE
Blog post for energy measurement of android devices

### PR DESCRIPTION
**Pull request by swoods-scottlogic**
Please add a direct link to your post here:

https://swoods-scottlogic.github.io/blog/2024/04/12/measuring-android-energy-use.html

Have you (please tick each box to show completion):

 - [X] Added your blog post to a single category?
 - [X] Added a brief summary for your post? Summaries should be roughly two sentences in length and give potential readers a good idea of the contents of your post.
 - [X] Checked that the build passes?
 - [X] Checked your spelling (you can use `npm spellcheck` if that's your thing)
 - [X] Ensured that your author profile contains a profile image, and a brief description of yourself? (make it more interesting than just your job title!)
 - [X] Optimised any images in your post? They should be less than 100KBytes as a general guide.

Posts are reviewed / approved by your Regional Tech Lead.

### CODE REVIEW

**Comment by swoods-scottlogic** [comment]
This also required  https://github.com/ScottLogic/blog/pull/186 to be merged as well do that Josh Isted can be properly credited on the contributors page.

File: _posts/2024-04-12-measuring-android-energy-use.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+---
+title: 'Measuring Energy use of Android Devices'
+date: 2024-04-23 00:00:00 Z
+categories:
+- Sustainability
+tags:
+- Sustainability
+summary: A post about how to gather power usage of android devices without the needs for external tools or equipment.
```
Comment (jwalker-scottlogic) [inline _posts/2024-04-12-measuring-android-energy-use.md]: Can we change `android` to `Android` here and the Introduction section to match other uses

File: _posts/2024-04-12-measuring-android-energy-use.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+author: swoods
+contributors: jisted
+---
+
+# Introduction
+As part of a project onto the carbon footprint of mobile computing (CFoMC), we required a method to be able to record the energy use of certain computational workloads on differing mobile devices, part of that being android mobile devices. So we needed an method to accurately measure the energy use on a device.
```
Comment (jwalker-scottlogic) [inline _posts/2024-04-12-measuring-android-energy-use.md]: `an method` -> `a method`

File: _posts/2024-04-12-measuring-android-energy-use.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+As part of a project onto the carbon footprint of mobile computing (CFoMC), we required a method to be able to record the energy use of certain computational workloads on differing mobile devices, part of that being android mobile devices. So we needed an method to accurately measure the energy use on a device. 
+
+## Why measure energy use?
+The reason why we wanted to measure the energy use was to compare the energy use of the same code / calculations across different devices. To do this we used some standard off-the-shelf mathematically CPU heavy tests to be able to keep repeatable results. 
+
+## Methods of measurement
+From investigation we determined there was many ways to measure energy use, generally split into three different methodologies.
+
+### Hardware Based
+Power meters / monitoring hardware can provide energy vales based on how much power the phone is drawing when fully charged to perform a particular process. This is the preferred method of several companies that perform energy evaluations of applications, but consumer-grade power monitors also claim to have an accuracy of +-2%.
```
Comment (jwalker-scottlogic) [inline _posts/2024-04-12-measuring-android-energy-use.md]: `vales` -> `values`

File: _posts/2024-04-12-measuring-android-energy-use.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+### Model Based
+This methodology takes indicative energy usage values for a given component and calculates the power used based on modelled data. For example, “Processor x running at frequency x for z amount of time uses this amount of energy.” These calculations are then repeated for each piece of equipment concerned and totalled up to an estimate. These methods do not typically give accurate numerical values, but are used to work out relative consumption. The Android Studio profiler seem to be based on this method.
```
Comment (jwalker-scottlogic) [inline _posts/2024-04-12-measuring-android-energy-use.md]: `profiler seem to be based` -> `profiler seems to be based`

Comment (jwalker-scottlogic) [inline _posts/2024-04-12-measuring-android-energy-use.md]: `frequency x` -> `frequency y` ?

File: _posts/2024-04-12-measuring-android-energy-use.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    currentInAmps = recordedCurrent / 1000000.0
+    joules = currentInAmps * timeDifference * voltage
+~~~~~~
+
+For our measurements we took samples every 1 second, so when calculated we had a array of joules used in each second of the process. To get the total amount of energy used we had to sum up all these values to a figure for the entire process. 
+
+![Example of Data Points]({{ site.github.url }}/swoods/assets/CFoMC_AndroidAppConsole.png)
+
+### Pitfalls and limitations
+There are a few limitations to measuring the devices power by these methods, most can be negated or controlled by ensure certain device conditions
```
Comment (jwalker-scottlogic) [inline _posts/2024-04-12-measuring-android-energy-use.md]: `ensure` -> `ensuring`

File: _posts/2024-04-12-measuring-android-energy-use.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+If there are other apps / system processes running they may impact results. While running tests for the CFoMC project we had a set list of conditions a device must be in for each test.
+
+- We use power to monitor the power
+
+Our data collection service is clearly going to be consuming some power when collecting energy values. This is inevitable using the background service approach, but it should be a consistent, low trickle of energy across all workloads. To mitigate this, we created a ‘control’ workload with a simple timer so we could consider baseline energy usage in our analysis.
+
+
+## Links / Sources etc
+- [Battery Manager API](https://developer.android.com/reference/kotlin/android/os/BatteryManager)
+- [Carbon Footprint of Mobile Computing GitHub](https://github.com/ScottLogic/Mobile-Carbon-Android)
```
Comment (jwalker-scottlogic) [inline _posts/2024-04-12-measuring-android-energy-use.md]: This gives a 404 for me, is it private? Need to check it's accessible when not in the ScottLogic organisation or remove the link

File: _posts/2024-04-12-measuring-android-energy-use.md (hunk 1 above)
Comment (swoods-scottlogic) [reply _posts/2024-04-12-measuring-android-energy-use.md]: Updated and checked every instance of `Android` to ensure using the same capitalisation.

File: _posts/2024-04-12-measuring-android-energy-use.md (hunk 6 above)
Comment (swoods-scottlogic) [reply _posts/2024-04-12-measuring-android-energy-use.md]: Removed link, as the git repo is private for SL only.

**Pull request merged**